### PR TITLE
docs: add Cask Cookbook

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Move docs from the cask repo to docs.brew.sh
CC @Homebrew/cask

### CODE REVIEW

File: docs/Cask-Cookbook.md
Comment (Rylan12) [inline docs/Cask-Cookbook.md:108]:
```suggestion
[`<cask-token>`](#token-reference) should match the Cask filename, without the `.rb` extension,
```

**Review by Rylan12 — APPROVED** [review]
I haven't gone through this in much detail as I know it's just adapted from the existing Cask documentation. I do plan on spending some time over in the cask repos this week or next week to help refine this.

Glad it's moving to the central docs repo, though!

**Comment by SMillerDev** [comment]
Yeah, there is probably a lot that could be better now that it will be centralized. But I figured we can move it first

**Comment by vitorgalvao** [comment]
This is a huge file, and though I may be biased I find it harder to follow than the previous solution. Important information is mixed with comparatively minor details, because this wasn’t designed to be consumed in one sitting.

These docs were made to:

1. Give an overview of the bare minimum one needs to know.
2. Link to specific documentation that goes from a simple explanation to in-depth reasoning.

They were loosely based on the Electron docs, which make it super easy and clear to find just what you need. Navigating to a specific document, especially via <kbd>t</kbd>, is simpler than having to <kbd>⌘</kbd><kbd>F</kbd> on a single document and get unrelated matches. That goes both for users searching and maintainers linking to specific sections.

I don’t want people to see a wall of text and give up; I want them to read one section and feel they have a good understanding of that area. They should get a feeling of accomplishment (and we’ve gotten praise on the doc’s structure for that). Single files are more conductive to that: at a glance you can see the complexity (or lack thereof) of a feature.

I’m not saying the single file solution is wrong (far from it); rather that the Cask docs were written to be consumed one way, and shoehorning them into a different structure doesn't work well.

If the one file solution is to be followed (for consistency with the formulae cookbook), I’d say this will need a table of contents at the top, at least until it’s rewritten. Unless you plan on having the rewrite finished before the PR is merged.

**Comment by SMillerDev** [comment]
> Navigating to a specific document, especially via t, is simpler than having to ⌘F on a single document and get unrelated matches. That goes both for users searching and maintainers linking to specific sections.

With Algolia DocSearch on docs.brew.sh it should be pretty easy to find what you're looking for. I don't think I ever use Cmd + F on there.

> They should get a feeling of accomplishment (and we’ve gotten praise on the doc’s structure for that).

I'm curious how many people praised the current way and how many silently gave up on it (like I did on my first PR) and just copied other Casks until it worked.

> If the one file solution is to be followed (for consistency with the formulae cookbook), I’d say this will need a table of contents at the top, at least until it’s rewritten. Unless you plan on having the rewrite finished before the PR is merged.

I don't plan to rewrite anything here, There is a list of stanzas with links pretty close to the top though, I figured that would work pretty well as a table of contents.

**Comment by vitorgalvao** [comment]
> With Algolia DocSearch on docs.brew.sh it should be pretty easy to find what you're looking for. I don't think I ever use Cmd + F on there.

As a counterpoint, I never use the website search (on either Homebrew or other projects) and always go for the GitHub docs.

I doubt I’m alone, but I won’t argue that point. I don’t have the data to say which is more popular, and as I said, I understand the need for consistency.

> There is a list of stanzas with links pretty close to the top though, I figured that would work pretty well as a table of contents.

But there are non-stanza-related sections in there, like the token reference.

> I'm curious how many people praised the current way and how many silently gave up on it (like I did on my first PR) and just copied other Casks until it worked.

I don’t see how that’s either fair or constructive.

For one, we can’t know what’s wrong unless people tell us, so all we have to go on is the ratio of people who praise VS criticise. And as you should be well aware by now, users tend to criticise way more than praise.

For another, one could use your exact words to describe the one-document wall-of-text approach, or literally any other.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by SMillerDev** [comment]
> But there are non-stanza-related sections in there, like the token reference.

We can move that one up between the required/optional tables and the actual stanza explanations. That way people can spot it quickly.

I'm not in favor of a ToC because it'll have to be generated (and won't show up in the github repo) or it'll have to be manually written (and become outdated pretty easily)

**Comment by MikeMcQuaid** [comment]
> Edit: can I suggest we merge this as is (but move the token reference) and work on improving it later?

👍🏻 from me.

**Comment by SMillerDev** [comment]
I didn't move anything yet?

**Comment by vitorgalvao** [comment]
> I didn't move anything yet?

It can be done in subsequent PRs. If it’s going to be improved over time, it’s easier to see the changes in place than track modifications to a huge document. Especially since the rich diff will be closer to how it really looks.

“Spotting the token reference quickly” wasn’t the issue; it’s different levels and types of information strung together in an inappropriate structure to what it was designed for. Moving one section around won’t have a big impact.

If the future is one giant document, it’s more productive to merge it soon to detect problems in a real context. Plus, if the Algolia search is the main supported way to reference our docs, I’d like to experiment with it to make more informed suggestions.

**Pull request merged**
